### PR TITLE
[ci] #4151: Enhance the test environment by automatically generating WASM files for the smartcontracts to be tested

### LINKED ISSUE #4151: Enhance `test_env.py` to Automate WASM Smart Contract Generation and Organization
### Feature Request
We propose an enhancement to the `test_env.py` script used in setting up the Iroha test environment. The enhancement involves automating the generation of WebAssembly (WASM) files for smart contracts and organizing these files into a designated directory (`smartcontracts`) within the `test` directory.

### Motivation
The current process of managing smart contracts in the Iroha test environment requires manual generation and organization of WASM files. This is not only time-consuming but also prone to human error. By automating this process, we can ensure consistency, reduce manual effort, and make it easier to manage and add new smart contracts in the future.

### PR DESCRIPTION
**Pull request by Asem-Abdelhady**
## Description

I made a shell script to create test-smartcontracts directory and build the WASM files in that directory
### Linked issue

Closes #4151 

### Benefits

No need to manually generate the WASM files.

### Checklist

- [X] I've read `CONTRIBUTING.md`
- [X] I've used the standard signed-off commit format (or will squash just before merging)
- [X] All applicable CI checks pass (or I promised to make them pass later)
- [ ] I replied to all comments after code review, marking all implemented changes with thumbs up

### CODE REVIEW

**Comment by Erigara** [comment]
Please only add relevant commits

File: scripts/test_env.py [hunk 1]
```diff
@@ -143,6 +143,35 @@ def run(self, shared_env: dict(), submit_genesis: bool = False):
...
+def initialize_test_smart_contracts(root_dir: pathlib.Path, out_dir: pathlib.Path):
+    """
+    Function to make directory called smartcontracts, build every integeration smartcontract,
+    the it moves the result WASM file to the smartcontracts directory. 
+    """
+    smartcontracts_dir = out_dir / "smartcontracts"
+    os.makedirs(smartcontracts_dir, exist_ok=True)
+    source_dir = pathlib.Path("./client/tests/integration/smartcontracts")
+
+    #Iterate through every smartcontract in the inte
```
[the remark this answers is not in the document]
Comment (Asem-Abdelhady) [reply scripts/test_env.py]: was typo somehow the full comment was not committed.

File: scripts/test_env.py [hunk 2]
```diff
@@ -143,6 +143,35 @@ def run(self, shared_env: dict(), submit_genesis: bool = False):
...
+        wasm_file_name = f"{folder_name}.wasm"
+        wasm_file_path = root_dir / wasm_file_name
+
+        # Build the smart contracts in the root director
+        subprocess.run([
+            "cargo", "run", "--bin", "iroha_wasm_builder_cli", "--", "build",
+            str(folder), "--optimize", "--outfile", str(wasm_file_path)
+        ])
+
+        # Move the .wasm file to the smarts contracts folder
```
[the remark this answers is not in the document]
Comment (Asem-Abdelhady) [reply scripts/test_env.py]: Done

File: scripts/test_env.py [hunk 3]
```diff
@@ -143,6 +143,31 @@ def run(self, shared_env: dict(), submit_genesis: bool = False):
...
+def initialize_test_smart_contracts(out_dir: pathlib.Path):
+    """
+    Function to make directory called smartcontracts, build every integeration smartcontract,
+    the it moves the result WASM file to the smartcontracts directory. 
+    """
+    smartcontracts_dir = out_dir / "smartcontracts"
+    os.makedirs(smartcontracts_dir, exist_ok=True)
+    source_dir = pathlib.Path("./client/tests/integration/smartcontracts")
+
+    #Iterate through every smartcontract in the integeration tests folder
```
Comment (Erigara) [inline scripts/test_env.py:155]:
```suggestion
    #Iterate through every smartcontract in the integration tests folder
```

**Review by Erigara — COMMENTED** [review]
I was thinking about should we make smartcontract build optional?

**Comment by Asem-Abdelhady** [comment]
> I was thinking about should we make smartcontract build optional?

@Erigara We can add argument `--wasm_generation` with default true value. We check the arg before the generation step.

**Comment by Erigara** [comment]
I suggested to make this optional is the fact that this functionality is only necessary to run python integration tests. 
So if i don't need wasm files i could skip this step and don't waste my time.

File: scripts/test_env.py [hunk 4]
```diff
@@ -143,6 +143,31 @@ def run(self, shared_env: dict(), submit_genesis: bool = False):
...
+def initialize_test_smart_contracts(out_dir: pathlib.Path):
+    """
+    Function to make directory called smartcontracts, build every integeration smartcontract,
+    the it moves the result WASM file to the smartcontracts directory. 
+    """
+    smartcontracts_dir = out_dir / "smartcontracts"
+    os.makedirs(smartcontracts_dir, exist_ok=True)
+    source_dir = pathlib.Path("./client/tests/integration/smartcontracts")
+
+    #Iterate through every smartcontract in the integration tests directory
```
Comment (mversic) [inline scripts/test_env.py:155]:
```suggestion
```

useless

**Comment by mversic** [comment]
> I suggested to make this optional is the fact that this functionality is only necessary to run python integration tests.
So if i don't need wasm files i could skip this step and don't waste my time.

I don't think we should put the test smart contract build into `test_env` script at all

**Comment by Asem-Abdelhady** [comment]
> > I suggested to make this optional is the fact that this functionality is only necessary to run python integration tests.
> > So if i don't need wasm files i could skip this step and don't waste my time.
> 
> I don't think we should put the test smart contract build into `test_env` script at all

Do you think it is better to add it in the CI process? or not to do it at all?

File: scripts/test_env.py (hunk 4 above)
Comment (Asem-Abdelhady) [reply scripts/test_env.py]: Done

**Comment by Asem-Abdelhady** [comment]
@AlexStroke  We reached an approach to make a separate script and add it in the CI process so those who want to set up the test env don't have to build the smartcontracts. The script will be done using Github artifacts. I will search about how exactly it works then I will commit in the PR accordingly. Is that okay with you?

**Comment by AlexStroke** [comment]
> @AlexStroke We reached an approach to make a separate script and add it in the CI process so those who want to set up the test env don't have to build the smartcontracts. The script will be done using Github artifacts. I will search about how exactly it works then I will commit in the PR accordingly. Is that okay with you?

yes

**Comment by Asem-Abdelhady** [comment]
> > @AlexStroke We reached an approach to make a separate script and add it in the CI process so those who want to set up the test env don't have to build the smartcontracts. The script will be done using Github artifacts. I will search about how exactly it works then I will commit in the PR accordingly. Is that okay with you?
> 
> yes

Github artifacts are not needed anymore. The shell script provided does the job, you can use it by passing the SOURCE-DIR. The script can be integerated into the CI as well.

File: .gitignore [hunk 5]
```diff
@@ -49,3 +49,4 @@ result
 /test/
 /iroha-java/
 /lcov.info
+**/test-smartcontracts/
```
Comment (0x009922) [inline .gitignore:52]: Quite... indeterminate. I guess the target is always at the root?
```suggestion
/test-smartcontracts
```

**Review by 0x009922 — CHANGES_REQUESTED** [review]
Please also document somewhere how to run the script, so that a person could simply cope-paste the command into shell.

File: .gitignore (hunk 5 above)
Comment (Asem-Abdelhady) [reply .gitignore:52]: it depends on from where you are running the script. if you are running from root then it will be in root yes

**Comment by Stukalov-A-M** [comment]
@Asem-Abdelhady, the Iroha-java team is already created script for autogenerating WASMs, I think you may just reuse it.
https://github.com/hyperledger/iroha-java/blob/iroha2-dev/modules/client/src/test/executor/build.sh

**Comment by 0x009922** [comment]
@Stukalov-A-M, IMO both scripts are quite small and trivial, and it doesn't seem like there is something to reuse really.

**Comment by Asem-Abdelhady** [comment]
> @Stukalov-A-M, IMO both scripts are quite small and trivial, and it doesn't seem like there is something to reuse really.

i agree

File: .gitignore (hunk 5 above)
Comment (Asem-Abdelhady) [reply .gitignore:52]: I believe it is better to leave it this way. maybe someone wanted to create a dir locally to test something so it makes sense if he ran the command within this dir to have test-smartcontracts inside that that dir as well.

File: scripts/generate_wasm.sh [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    fi
+done
+
+echo "Smart contracts build complete."
+
+# How to run:
+# make sure you root from the root directrory if iroha or
+# any iroha directory that is not a package in the working space
+# run the following command:
+# ./path/to/script/generate_wasm.sh /path/to/smart_contracts
```
[the remark this answers is not in the document]
Comment (Asem-Abdelhady) [reply scripts/generate_wasm.sh]: Done

File: scripts/generate_wasm.sh [hunk 7]
```diff
@@ -0,0 +1,28 @@
+#!/bin/bash
```
Comment (Erigara) [inline scripts/generate_wasm.sh:1]: If you don't use any `bash` specific functionality i would suggest to use `#!/bin/sh`

Comment (Asem-Abdelhady) [reply scripts/generate_wasm.sh]: done

**Review by AlexStroke — APPROVED** [review]
LGTM

**Review by Erigara — APPROVED** [review]
LGTM

**Review by 0x009922 — APPROVED** [review]

**Review by 0x009922 — APPROVED** [review]

**Review by Erigara — APPROVED** [review]

**Pull request merged**